File: src/wings/skills/injector.py
"""Skill injector — adds available skills to the system prompt."""

from __future__ import annotations

from xml.sax.saxutils import escape as xml_escape

from wings.skills.types import SkillSpec


class SkillInjector:
    """Injects available skill names and descriptions into the system prompt.

    The model sees an <available_skills> XML block it can use to
    decide whether to call skill_view() for a particular skill.
    """

    def inject_skills(
        self,
        system_prompt: str,
        skills: list[SkillSpec],
    ) -> str:
        """Append an <available_skills> block to the system prompt."""
        visible = [s for s in skills if not s.disable_model_invocation]
        if not visible:
            return system_prompt

        block = self._build_skills_block(visible)
        return f"{system_prompt}\n\n{block}"

    def _build_skills_block(self, skills: list[SkillSpec]) -> str:
        lines = [
            "## Skills",
            "Skills are optional task playbooks. Use them only when a listed entry",
            "clearly matches the user's current request.",
            "Call skill_view(name=\"<skill_name>\") to load a skill's full instructions,",
            "then use only the tools available in this session.",
            "When no entry is relevant, answer without loading a skill.",
            "",
            "<available_skills>",
        ]
        for s in skills:
            lines.append("  <skill>")
            lines.append(f"    <name>{xml_escape(s.name)}</name>")
            lines.append(f"    <description>{xml_escape(s.description)}</description>")
            lines.append("  </skill>")
        lines.append("</available_skills>")
        return "\n".join(lines)

File: src/skills/injector.ts
/**
 * Skill injector — adds available skills to the system prompt.
 *
 * Handles both bundled skills (TypeScript) and file-based skills (SKILL.md).
 */

import type { SkillSpec } from "./types.ts";
import { getBundledSkills, type BundledSkillDefinition } from "./bundledSkills.ts";

function xmlEscape(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

export class SkillInjector {
  /** Append an <available_skills> block to the system prompt.
   *  Includes both bundled skills and file-based skills. */
  injectSkills(systemPrompt: string, fileSkills: SkillSpec[]): string {
    const bundled = getBundledSkills().filter((s) => !s.disableModelInvocation);
    const fileVisible = fileSkills.filter((s) => !s.disable_model_invocation);
    if (bundled.length === 0 && fileVisible.length === 0) return systemPrompt;
    return `${systemPrompt}\n\n${this._buildBlock(bundled, fileVisible)}`;
  }

  private _buildBlock(
    bundled: BundledSkillDefinition[],
    fileSkills: SkillSpec[],
  ): string {
    const lines: string[] = [
      "## Skills",
      "Skills are optional task playbooks. Use them only when a listed entry",
      "clearly matches the user's current request.",
      'Call skill_view(name="<skill_name>") to load a skill\'s full instructions,',
      "then use only the tools available in this session.",
      "When no entry is relevant, answer without loading a skill.",
      "",
      "<available_skills>",
    ];
    for (const s of bundled) {
      lines.push("  <skill>");
      lines.push(`    <name>${xmlEscape(s.name)}</name>`);
      lines.push(`    <description>${xmlEscape(s.description)}</description>`);
      lines.push("  </skill>");
    }
    for (const s of fileSkills) {
      lines.push("  <skill>");
      lines.push(`    <name>${xmlEscape(s.name)}</name>`);
      lines.push(`    <description>${xmlEscape(s.description)}</description>`);
      lines.push("  </skill>");
    }
    lines.push("</available_skills>");
    return lines.join("\n");
  }
}

